Framework/Algorithms/src/RebinToWorkspace.cpp

```diff
@@ -43,55 +49,162 @@ void RebinToWorkspace::init() {
+      convert->setProperty<MatrixWorkspace_sptr>("OutputWorkspace", outputWS);
+      convert->setChildStartProgress(0.);
+      convert->setChildStartProgress(endProgress);
+      convert->executeAsChildAlg();
+      outputWS = convert->getProperty("OutputWorkspace");
+    } else {
+      outputWS = inputWS->clone();
+    }
+  }
+
+  return outputWS;
+}
+
+// this follows closely what Rebin does except each x-axis is different
+void RebinToWorkspace::rebin(MatrixWorkspace_sptr &toRebin,
+                             MatrixWorkspace_sptr &toMatch) {
+  // create the output workspace
+  auto outputWS = this->createOutputWorkspace(toRebin, .5);
+  auto outputWSEvents = boost::dynamic_pointer_cast<EventWorkspace>(outputWS);
+
+  const int numHist = static_cast<int>(toRebin->getNumberHistograms());
+  Progress prog(this, 0.5, 1.0, numHist);
+
+  // everything gets the same bin boundaries
+  const bool matchingX = (toMatch->getNumberHistograms() == 1);
+
+  // rebin
+  PARALLEL_FOR_IF(Kernel::threadSafe(*toMatch, *outputWS))
+  for (int i = 0; i < numHist; ++i) {
+    PARALLEL_START_INTERUPT_REGION
+    const auto &edges = matchingX ? toMatch->histogram(0).binEdges()
+                                  : toMatch->histogram(i).binEdges();
+    if (m_isEvents) {
+      outputWSEvents->getSpectrum(i).setHistogram(edges);
+    } else {
+      outputWS->setHistogram(
+          i, HistogramData::rebin(toRebin->histogram(i), edges));
+    }
+    prog.report();
+    PARALLEL_END_INTERUPT_REGION
+  }
+  PARALLEL_CHECK_INTERUPT_REGION
+
+  setProperty("OutputWorkspace", outputWS);
+}
+
+// handles the special case when binning from EventWorkspace to Workspace2D most
+// efficiently by histogramming the events to the correct binning directly
+void RebinToWorkspace::histogram(API::MatrixWorkspace_sptr &toRebin,
+                                 API::MatrixWorkspace_sptr &toMatch) {
+  const auto &inputWS =
+      boost::dynamic_pointer_cast<const EventWorkspace>(toRebin);
+  const int numHist = static_cast<int>(toRebin->getNumberHistograms());
+  // everything gets the same bin boundaries
+  const bool matchingX = (toMatch->getNumberHistograms() == 1);
+
+  auto outputWS = DataObjects::create<API::HistoWorkspace>(*toRebin);
+  Progress prog(this, 0.25, 1.0, numHist);
+
+  // histogram
+  PARALLEL_FOR_IF(Kernel::threadSafe(*toMatch, *outputWS))
+  for (int i = 0; i < numHist; ++i) {
+    PARALLEL_START_INTERUPT_REGION
+    const auto &edges = matchingX ? toMatch->histogram(0).binEdges()
+                                  : toMatch->histogram(i).binEdges();
+
+    // TODO this should be in HistogramData/Rebin
+    const auto &eventlist = inputWS->getSpectrum(i);
+    MantidVec y_data(edges.size() - 1), e_data(edges.size() - 1);
+    eventlist.generateHistogram(edges.rawData(), y_data, e_data);
+
+    auto histo = HistogramData::Histogram(edges);
+    histo.setCounts(y_data);
+    histo.setCountStandardDeviations(e_data);
+
+    outputWS->setHistogram(i, histo);
```

Comment:
To avoid unnecessary copies and allocations: Move `y_data` and `e_data`.
You can also pass things directly to `setHistogram`, it is templated:

```cpp
outputWs->setHistogram(i, edges,
    Counts(std::move(y_data)),
    CountStandardDeviations(std::move(e_data)));
```